Add Gloas consensus type block package

terencechain opened the pull request.

terencechain committed:
Add Gloas protobuf definitions with spec tests

terencechain committed:
Implement Gloas fork support in consensus-types/blocks

james-prysm commented on proto/engine/v1/gloas.pb.go:
```diff
@@ -0,0 +1,288 @@
+	_ = protoimpl.EnforceVersion(protoimpl.MaxVersion - 20)
+)
+
+type ExecutionPayloadEnvelope struct {
```
hmm why was this file included without the protobuf file?

terencechain replied:
nice catch, the file was a left over and i checked it in by accident
in the latest commit i have that removed and addressed your other feedback

james-prysm commented on proto/prysm/v1alpha1/cloners.go:
```diff
@@ -61,3 +61,115 @@ func CopySyncCommitteeContribution(c *SyncCommitteeContribution) *SyncCommitteeC
+		copied.SyncAggregate = body.SyncAggregate.Copy()
+	}
+
+	copied.SignedExecutionPayloadBid = copySignedExecutionPayloadBid(body.SignedExecutionPayloadBid)
```
nit can we move this to the bottom since it's new?

james-prysm reviewed (dismissed):
LGTM

terencechain commented on beacon-chain/blockchain/process_block_test.go:
```diff
@@ -2805,6 +2805,10 @@ func TestProcessLightClientUpdate(t *testing.T) {
 	require.NoError(t, s.cfg.BeaconDB.SaveHeadBlockRoot(ctx, [32]byte{1, 2}))
 
 	for _, testVersion := range version.All()[1:] {
+		if testVersion == version.Gloas {
+			// TODO(16027): Unskip light client tests for Gloas
+			continue
+		}
```
note we are going to redo this, we are experimenting a different approach in: https://github.com/OffchainLabs/prysm/pull/16030
but it shouldn't block anyone reviewing this PR

potuz commented on consensus-types/blocks/factory.go:
```diff
@@ -82,13 +82,15 @@ func NewSignedBeaconBlock(i any) (interfaces.SignedBeaconBlock, error) {
 		return initBlindedSignedBlockFromProtoFulu(b)
 	case *eth.GenericSignedBeaconBlock_BlindedFulu:
 		return initBlindedSignedBlockFromProtoFulu(b.BlindedFulu)
+	case *eth.SignedBeaconBlockGloas:
```
Do we need a "Generic" version like for previous forks, something like `b.Gloas.Block`?

potuz commented on consensus-types/blocks/factory.go:
```diff
@@ -138,13 +140,15 @@ func NewBeaconBlock(i any) (interfaces.ReadOnlyBeaconBlock, error) {
 		return initBlindedBlockFromProtoFulu(b)
 	case *eth.GenericBeaconBlock_BlindedFulu:
 		return initBlindedBlockFromProtoFulu(b.BlindedFulu)
+	case *eth.BeaconBlockGloas:
```
Do we need a generic version like previous forks? `b.Gloas.Block`

potuz commented on consensus-types/blocks/factory.go:
```diff
@@ -629,6 +641,8 @@ func BuildSignedBeaconBlockFromExecutionPayload(blk interfaces.ReadOnlySignedBea
 			},
 			Signature: sig[:],
 		}
+	case version.Gloas:
```
This case should fail much earlier when trying to get the Execution from the body. It's better to return early from this function with an error. Also, the fact that this function passed tests shows that this couldn't have been covered. So a regression test is needed.

potuz commented on consensus-types/blocks/getters.go:
```diff
@@ -157,14 +159,17 @@ func (b *SignedBeaconBlock) PbGenericBlock() (*eth.GenericSignedBeaconBlock, err
 			Block: &eth.GenericSignedBeaconBlock_Fulu{Fulu: bc},
 		}, nil
+	case version.Gloas:
+		// Gloas doesn't support GenericSignedBeaconBlock yet
```
This answers my pervious questions.... Probably better to discuss this offline

potuz commented on consensus-types/blocks/getters.go:
```diff
@@ -666,6 +677,16 @@ func (b *SignedBeaconBlock) UnmarshalSSZ(buf []byte) error {
+		if err := pb.UnmarshalSSZ(buf); err != nil {
+			return err
+		}
+		var err error
```
I know this is the design on all previous cases, but is there any reason to not have done:

```suggestion
		err := pb.UnmarshalSSZ(buf); 
		if err != nil {
			return err
		}
```

potuz commented on consensus-types/blocks/proto.go:
```diff
@@ -1477,3 +1521,67 @@ func initBlindedBlockBodyFromProtoFulu(pb *eth.BlindedBeaconBlockBodyElectra) (*
+		voluntaryExits:            pb.VoluntaryExits,
+		syncAggregate:             pb.SyncAggregate,
+		signedExecutionPayloadBid: pb.SignedExecutionPayloadBid,
+		blsToExecutionChanges:     pb.BlsToExecutionChanges,
```
Why the ordering?

potuz commented on consensus-types/blocks/setters.go:
```diff
@@ -182,3 +182,21 @@ func (b *SignedBeaconBlock) SetExecutionRequests(req *enginev1.ExecutionRequests
 	b.block.body.executionRequests = req
 	return nil
 }
+
```
`SetExecutionRequests` should also fail if the version is Gloas or more, same for the Execution Payload setter.

potuz commented on consensus-types/blocks/getters.go:
```diff
@@ -80,6 +80,8 @@ func (b *SignedBeaconBlock) Copy() (interfaces.SignedBeaconBlock, error) {
 			return initBlindedSignedBlockFromProtoFulu(pb.(*eth.SignedBlindedBeaconBlockFulu).Copy())
 		}
 		return initSignedBlockFromProtoFulu(pb.(*eth.SignedBeaconBlockFulu).Copy())
+	case version.Gloas:
```
Not related to this line. But the getters file needs to be futher modified. Some getters do not make sense for Gloas and instead of returning nil and risking a panic they should return an error. For example the current path for `Execution()` will return `nil` instead of erroring out. The same for the `ExecutionRequests()`. We should add errors for all getters that would fail.

potuz commented on consensus-types/blocks/types.go:
```diff
@@ -40,23 +40,25 @@ var (
+	blobKzgCommitments        [][]byte
+	executionRequests         *enginev1.ExecutionRequests
+	payloadAttestations       []*eth.PayloadAttestation
+	signedExecutionPayloadBid *eth.SignedExecutionPayloadBid
```
Reverse order here.

potuz reviewed (commented):
A bunch of getters should now error on Gloas, at least Execution, KZG commitments, Execution Requests. 

First serious pass, will do another in a couple of hours.

terencechain committed:
Potuz's feedback

terencechain committed:
Go fmt

terencechain committed:
Better conditions for Execution()

terencechain committed:
IsExecutionEnabled for post capella state

terencechain committed:
Merge branch 'develop' into add-gloas-block

Inspector-Butters committed:
Refactor slot by blockroot (#16040)

**Review after #16033 is merged**

**What type of PR is this?**
Other

**What does this PR do? Why is it needed?**
This PR refactors the code to find the corresponding slot of a given
block root using state summary or the block itself, into its own
function `SlotByBlockRoot(ctx, blockroot)`.

Note that there exists a function `slotByBlockRoot(ctx context.Context,
tx *bolt.Tx, blockRoot []byte)` immediately below the new function. Also
note that this function has two drawbacks, which led to creation of the
new function:
- the old function requires a boltdb tx, which is not necessarily
available to the caller.
- the old function does NOT make use of the state summary cache. 

edit: 
- the old function also uses the state bucket to retrieve the state and
it's slot. this is not something we want in the state diff feature,
since there is no state bucket.

terencechain committed:
Check exec enabled condition to capella

terencechain committed:
Revert test want to false for empty bellatrix case
